File: packages/pcb/simulation/cross-validator.ts
/**
 * Multi-Physics Cross-Validation Engine.
 * Compares fast PINN surrogate predictions against reference numerical solvers.
 */

import { SolverResult, PhysicalMetric } from "./solvers";

export type CrossValidationStatus = "PASS" | "FAIL" | "WARNING" | "UNKNOWN";

export interface MetricComparison {
  metricName: string;
  referenceValue: number;
  surrogateValue: number;
  unit: string;
  absoluteDiscrepancy: number;
  relativeDiscrepancy: number; // e.g. 0.04 = 4%
  status: CrossValidationStatus;
}

export interface CrossValidationReport {
  overallStatus: CrossValidationStatus;
  comparisons: MetricComparison[];
  mae: number;
  rmse: number;
  maxRelativeDiscrepancy: number;
  evaluatedAt: number;
}

export class SimulationCrossValidator {
  public static validate(
    referenceResults: SolverResult[],
    surrogateResult: SolverResult
  ): CrossValidationReport {
    const comparisons: MetricComparison[] = [];
    const surrogateMap = new Map<string, PhysicalMetric>();

    for (const m of surrogateResult.metrics) {
      surrogateMap.set(m.name, m);
    }

    let sumAbsError = 0.0;
    let sumSqError = 0.0;
    let maxRelError = 0.0;
    let validPairs = 0;

    for (const refRes of referenceResults) {
      for (const refMetric of refRes.metrics) {
        const surMetric = surrogateMap.get(refMetric.name);
        if (!surMetric) continue;

        const absDiff = Math.abs(surMetric.value - refMetric.value);
        const relDiff = refMetric.value !== 0 ? absDiff / Math.abs(refMetric.value) : absDiff;

        sumAbsError += absDiff;
        sumSqError += absDiff * absDiff;
        maxRelError = Math.max(maxRelError, relDiff);
        validPairs++;

        let status: CrossValidationStatus = "PASS";
        if (relDiff > 0.15) {
          status = "FAIL";
        } else if (relDiff > 0.05) {
          status = "WARNING";
        }

        comparisons.push({
          metricName: refMetric.name,
          referenceValue: refMetric.value,
          surrogateValue: surMetric.value,
          unit: refMetric.unit,
          absoluteDiscrepancy: Number(absDiff.toFixed(3)),
          relativeDiscrepancy: Number(relDiff.toFixed(4)),
          status,
        });
      }
    }

    if (validPairs === 0) {
      return {
        overallStatus: "UNKNOWN",
        comparisons: [],
        mae: 0,
        rmse: 0,
        maxRelativeDiscrepancy: 0,
        evaluatedAt: Date.now(),
      };
    }

    const mae = Number((sumAbsError / validPairs).toFixed(3));
    const rmse = Number(Math.sqrt(sumSqError / validPairs).toFixed(3));

    let overallStatus: CrossValidationStatus = "PASS";
    if (comparisons.some((c) => c.status === "FAIL")) {
      overallStatus = "FAIL";
    } else if (comparisons.some((c) => c.status === "WARNING")) {
      overallStatus = "WARNING";
    }

    return {
      overallStatus,
      comparisons,
      mae,
      rmse,
      maxRelativeDiscrepancy: Number(maxRelError.toFixed(4)),
      evaluatedAt: Date.now(),
    };
  }
}

File: backend/workline/pcb/simulation/cross_validator.py
"""Cross-validation between PINN surrogate predictions and reference numerical solvers."""

import math
import time
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field
from backend.workline.pcb.simulation.solvers import PhysicalMetric, SolverResult


class CrossValidationStatus(str):
    PASS = "PASS"
    FAIL = "FAIL"
    WARNING = "WARNING"
    UNKNOWN = "UNKNOWN"


class MetricComparison(BaseModel):
    metric_name: str
    reference_value: float
    surrogate_value: float
    unit: str
    absolute_discrepancy: float
    relative_discrepancy: float
    status: str  # PASS, FAIL, WARNING, UNKNOWN


class CrossValidationReport(BaseModel):
    overall_status: str
    comparisons: List[MetricComparison] = Field(default_factory=list)
    mae: float = 0.0
    rmse: float = 0.0
    max_relative_discrepancy: float = 0.0
    evaluated_at: float = Field(default_factory=time.time)


class SimulationCrossValidator:
    """Evaluates surrogate model fidelity against reference simulation outputs."""

    @classmethod
    def validate(
        cls, reference_results: List[SolverResult], surrogate_result: SolverResult
    ) -> CrossValidationReport:
        comparisons: List[MetricComparison] = []
        surrogate_map: Dict[str, PhysicalMetric] = {m.name: m for m in surrogate_result.metrics}

        sum_abs_err = 0.0
        sum_sq_err = 0.0
        max_rel_err = 0.0
        valid_pairs = 0

        for ref_res in reference_results:
            for ref_m in ref_res.metrics:
                sur_m = surrogate_map.get(ref_m.name)
                if not sur_m:
                    continue

                abs_diff = abs(sur_m.value - ref_m.value)
                rel_diff = abs_diff / abs(ref_m.value) if ref_m.value != 0 else abs_diff

                sum_abs_err += abs_diff
                sum_sq_err += abs_diff ** 2
                max_rel_err = max(max_rel_err, rel_diff)
                valid_pairs += 1

                status = "PASS"
                if rel_diff > 0.15:
                    status = "FAIL"
                elif rel_diff > 0.05:
                    status = "WARNING"

                comparisons.append(
                    MetricComparison(
                        metric_name=ref_m.name,
                        reference_value=ref_m.value,
                        surrogate_value=sur_m.value,
                        unit=ref_m.unit,
                        absolute_discrepancy=round(abs_diff, 3),
                        relative_discrepancy=round(rel_diff, 4),
                        status=status,
                    )
                )

        if valid_pairs == 0:
            return CrossValidationReport(
                overall_status="UNKNOWN",
                comparisons=[],
                mae=0.0,
                rmse=0.0,
                max_relative_discrepancy=0.0,
            )

        mae = round(sum_abs_err / valid_pairs, 3)
        rmse = round(math.sqrt(sum_sq_err / valid_pairs), 3)

        overall_status = "PASS"
        if any(c.status == "FAIL" for c in comparisons):
            overall_status = "FAIL"
        elif any(c.status == "WARNING" for c in comparisons):
            overall_status = "WARNING"

        return CrossValidationReport(
            overall_status=overall_status,
            comparisons=comparisons,
            mae=mae,
            rmse=rmse,
            max_relative_discrepancy=round(max_rel_err, 4),
        )
